Tests refacto

xakraz opened the pull request:
Refacto the tests sections:

- [x] Use docker-compose
- [x] Use BATS
- [x] Add Travis tests
- [x] Release on docker hub
```

xakraz committed:
[Misc] Moved Dockerfiles and corresponding files to specific subdirectories

xakraz committed:
[Tests] Refacto using docker-compose

xakraz commented:
@aladjadj I might need your feedback to not break the Docker build on docker hub

guilhem commented on client/cassh_docker.sh:
```diff
@@ -7,4 +7,4 @@ docker run --rm -it \
   --volume=${HOME}/.ssh:${HOME}/.ssh \
   --workdir=${HOME} \
-  nbeguier/cassh-client "$@"
+  leboncoin/cassh-client "$@"
```
looks to be [`leboncoin/cassh`](https://hub.docker.com/r/leboncoin/cassh/)

xakraz replied:
Yeah ... maybe we will bee `cassh` for the client and `cassh-server` for the server.

guilhem commented on tests/docker-compose.yml:
```diff
@@ -0,0 +1,29 @@
+    build:
+      context: ../server
+      dockerfile: Dockerfile
+    image: nbeguier/cassh-server
```
~nbeguier~ leboncoin?

xakraz replied:
We could but might be hard to push upstream

vdemonchy commented on client/Dockerfile:
```diff
@@ -1,16 +1,17 @@
 FROM alpine:latest
```
OPT: could/should be `FROM python:x.y-alpine`

xakraz replied:
I though the same. I would to it later

xakraz committed:
[CI] Splitted 'requirements.txt' resources and updated Travis config

xakraz committed:
[CI] Added e2e tests in Travis

xakraz committed:
[CI] Switch over BATS testing framework

xakraz commented:
Example with BATS

```
bats tests.bats 
 ✓ /ping
 ✓ /health
 ✓ Status unknown user
 ✓ Add user without username
 ✓ Add user with bad username
 ✓ Add user without realname
 ✓ Add user with no pubkey
 ✓ Add user with bad pubkey
 ✓ Add user
 ✓ add user named 'all' (should fail)
 ✓ status pending user
 ✓ add user with same realname (which is possible)
 ✓ updating user
 ✓ add user with same username (should fail)
 ✓ signing key without username
 ✓ signing key without realname
 ✓ signing key with no pubkey
 ✓ signing key with bad pubkey
 ✓ signing key when wrong public key
 ✓ signing key when PENDING status
 ✗ admin revoke 'toto'
   (in test file tests.bats, line 112)
     `[ "${RESP}" == 'Revoke user=toto.' ]' failed
        * Create CASSH-server SSH Keys
   Generating public/private rsa key pair.
   Your identification has been saved in test-keys/id_rsa_ca.
   Your public key has been saved in test-keys/id_rsa_ca.pub.
   The key fingerprint is:
   SHA256:DN5PnYdiSfldoU7vp6yEodo0NugJTiVOFh7XOzpJ9js root@xkrantz-mbp
   The key's randomart image is:
   +---[RSA 4096]----+
   |               . |
   |       .   .  . .|
   |    o o . o  o  .|
   |   . = + o +o+.. |
   |    = = S * =.o. |
   |   + = = * + ..  |
   |    + = B o .  ..|
   |   o o BE+ . . ..|
   |    . + o.  ..o  |
   +----[SHA256]-----+
   removed 'test-keys/id_rsa_ca'
   removed 'test-keys/id_rsa_ca.pub'
   removed 'test-keys/revoked-keys'
   removed directory 'test-keys'
 ✓ admin verify 'toto' status
 ✓ signing key when revoked
 ✓ delete 'toto'
 ✓ admin active unknown user
 ✓ admin verify 'testuser' status
 ✓ admin active testuser
 ✓ admin re-active testuser
 ✗ signing key
   (in test file tests.bats, line 153)
     `[ ssh-keygen -L -f /tmp/test-cert >/dev/null 2>&1 ]' failed with status 2
        * Create CASSH-server SSH Keys
   Generating public/private rsa key pair.
   Your identification has been saved in test-keys/id_rsa_ca.
   Your public key has been saved in test-keys/id_rsa_ca.pub.
   The key fingerprint is:
   SHA256:QwnnMem5spNvPvp5Uvz2s5OOWkSzupVscb1eyZ2B53U root@xkrantz-mbp
   The key's randomart image is:
   +---[RSA 4096]----+
   |      . +.       |
   |       +.+       |
   |       .+.  o    |
   |       .o  . o.. |
   |        So  +..oE|
   |      . ..o+ ++ O|
   |       + ...*  Bo|
   |      + +..=o.= .|
   |      .B=+oo.o+= |
   +----[SHA256]-----+
   removed 'test-keys/id_rsa_ca'
   removed 'test-keys/id_rsa_ca.pub'
   removed 'test-keys/revoked-keys'
   removed directory 'test-keys'
   removed '/tmp/test-cert'
 ✓ delete 'testuser'

30 tests, 2 failures
```

nbeguier commented:
Good idea to refacto, I wanted to do that for a long time.
To merge it, I would like you to update the README.md.
Also, we should work together to test your tests and document them as well.

xakraz committed:
[CI] Travis - Test docker-compose usage

* docker-compose - Updated image name
* Updated README + Travis deps

xakraz commented:
`README`  updated + added `tests_client.bats`

guilhem commented:
Ça dormir

aladjadj commented:
Be careful with the dockerhub autobuild configuration
- leboncoin/cassh is the client
- leboncoin/cassh-server is the server

xakraz commented:
@aladjadj That is exactly what I was thinking about 13days ago when I pinged you :p

aladjadj commented:
Ok I configured dockerhub for your branch, please report this change to master and tags before merge

xakraz committed:
[CI] Added 'Invoke' task runner and updated Pylint step

xakraz committed:
FIX-ME - Deactivated Lint on cassh-server

xakraz committed:
[Tests] Updated tests for pull request #11 from nbeguier/1_5_0

xakraz committed:
[CI] Added section to deploy to Docker Hub

xakraz committed:
[contrib] CASSH-server docker wrapper
